### ISSUE TITLE
Add overwrite functionality

### ISSUE THREAD

**Issue by KANAjetzt** [GodotModding/godot-mod-loader#14]
- We might be able to use the **`load_resource_pack()`** with `replace_files` set to `true`, to easily overwrite for example textures.
- That would make a mod like [Brotato-Explosion-Mute](https://github.com/BrotatoMods/Brotato-Explosion-Mute) as easy as locating the image file you want to replace and adding a new one at the mirrored location in the mod zip.
( In this case `res://projectiles/rocket/explosion.png` )

- If not handled with care, this has potential for carnage

**Comment by KANAjetzt** [comment]
I successfully tested this with [Brotato-Explosion-Mute](https://github.com/BrotatoMods/Brotato-Explosion-Mute), by allowing file replacement. 
The complexity to achieve the desired result is greatly reduced.


[KANA-ExplosionMuteV2.zip](https://github.com/GodotModding/godot-mod-loader/files/10355992/KANA-ExplosionMuteV2.zip)

**Comment by ithinkandicode** [comment]
From my reply on Discord:

> What about a separate folder? Eg `mods-overrides`. That way it'd give them a standardised structure and would make it clear when a mod is overriding something in vanilla 
> Then we could just `run _load_mod_zips` twice, with the second call being passed an arg that says to use the "overrides" folder and use `true` for `load_resource_pack`

**Comment by KANAjetzt** [comment]
> Darkly77 - Another use case for replacements: You could do stuff like make a balance pack for Brotato. Eg people often say Plank is weak, so someone could release a mod that improves its scaling by replacing the vanilla file.

**Comment by KANAjetzt** [comment]
> From my reply on Discord:
> 
> > What about a separate folder? Eg `mods-overrides`. That way it'd give them a standardised structure and would make it clear when a mod is overriding something in vanilla
> > Then we could just `run _load_mod_zips` twice, with the second call being passed an arg that says to use the "overrides" folder and use `true` for `load_resource_pack`

---

Now that I'm working on it - we can't do a separate folder because the only way to *overwrite* is to mirror the file paths inside the mod zip.

```
yourmod.zip
├───.import
└───mods-unpacked
    └───Author-ModName
        ├───mod_main.gd
        └───manifest.json
└───projectiles
    └───rocket
        └───explosion.png
```

There is also no point in adding something like ``"is_overwrite": true`` to the *manifest.json* That is loaded after we need to set the replace flag for ``ProjectSettings.load_resource_pack()``

So the only way to send data in this stage to the mod loader is the name of the zip file?
That would mean we can't use arbitrary zip names any more.

**Comment by ithinkandicode** [comment]
*(copied from Discord:)*

Hey, I've been looking into this too! I realised that allowing full overrides would mean we're loading stuff from mod ZIPs before we get a chance to read their meta data. That means a mod could be published that doesn't even have a manifest

So my approach is to simply replace resources from within mod_main. That ensures mods stay within our current standards

**Comment by ithinkandicode** [comment]
Example mod with my approach. Changes the color of the logo. Has 3 different colors, the last one mentioned in the code is the one that gets used.

**mod_main.gd** *(simplified)*:

```gdscript
extends Node

func _init(modLoader = ModLoader):
	var new1 = preload("res://mods-unpacked/Darkly77-LogoReplacement/overrides/ui/menus/title_screen/ui_logo__blue.png")
	new1.take_over_path("res://ui/menus/title_screen/ui_logo.png")
```

Download Mod (for [ML v3.1.0](https://github.com/GodotModding/godot-mod-loader/releases/tag/v3.1.0)):

- [Darkly77-LogoReplacement--ML-3.1.0.zip](https://github.com/GodotModding/godot-mod-loader/files/10435007/Darkly77-LogoReplacement--ML-3.1.0.zip)

**Comment by KANAjetzt** [comment]
## 🇦 ) Mod Loader Handling - Restart game once ( if a overwrite mod is added / removed )
<details>
 <summary>Details</summary>

- Creating an ``overwrites.gd`` inside the mod loader can work
- By checking for an ``overwrites`` folder at the root of each mod folder
- If detected it gets all files inside the ``overwrites`` and appends the required code to the ``overwrites.gd`` file

```python
var overwrite_index = preload("res://mods-unpacked/ModAuthor-ModName/overrides/path/to/overwrite/resource")
overwrite_index.take_over_path("res://path/to/original/resource")
```

- That ``overwrites.gd`` script gets initialized at the end of the mod loaders ``_init()`` function

- Problem with this solution, is that the ``overwrites.gd`` file is created on the first run of the game.
To Allow the replacement resources to be **preloaded** the file needs to exist before the start.
That's why a game restart is required.

- Currently a game can only be closed via script not restarted, this will most likely change in Godot 3.6:
https://github.com/godotengine/godot/pull/66168
</details>



## 🇧 ) Let mod authors create the ``overwrites.gd`` at the root of the mod folder
<details>
 <summary>Details</summary>

- My current proposed solution on how to handle the overwrites
- Mod authors create the ``overwrites.gd`` script and ``overwrites`` folder in there mod folder:
![image](https://user-images.githubusercontent.com/41547570/213876012-6eb574e7-1b41-49c8-b214-c9a6ddd91411.png)
- We provide tooling to make the creation of the ``overwrites.gd`` script a one click thing
- People ho don't want to use extra tooling can just write the ``overwrites.gd`` them self

**Example Mod with this setup:**
[KANA-OverwritesTest.zip](https://github.com/GodotModding/godot-mod-loader/files/10472488/KANA-OverwritesTest.zip)

**Testing Godot Project**
[ModdingTest.zip](https://github.com/GodotModding/godot-mod-loader/files/10472559/ModdingTest.zip)

</details>


## 🇨 ) use ``--script run.gd``
<details>
 <summary>Details</summary>

- Using the ``--script`` cmd arg to run a script before the game starts
- Not tested, because with that setup we will need to unzip the mods in the ``run.gd``
- And we want to get rid of it anyway

</details>

**Comment by KANAjetzt** [comment]
Just to reference it here as well:

There is also the possibility to *reload* a scene with a new ``PackedScene``.

- For that I used ``get_tree().change_scene_to(PackedScene)`` in a script extension of the scene that is the *owner* of the overwritten resource.

- That allows to use just ``load`` to load the replacement resource
- ``load`` can take variable resource paths
- That potentially allows to add a util function to the mod loader for mod authors to use

- I didn't explorer this further because the ``preload`` solution currently seems way cleaner and doesn't require to *reload* / *replace* in a scenes ``_ready()`` function.

**Example Brotato Mod of this solution:**
[Darkly77-LogoReplacement.zip](https://github.com/GodotModding/godot-mod-loader/files/10472593/Darkly77-LogoReplacement.zip)

**Comment by KANAjetzt** [comment]
closed by https://github.com/GodotModding/godot-mod-loader/pull/74

**Comment by ithinkandicode** [comment]
> Darkly77 - Another use case for replacements: You could do stuff like make a balance pack for Brotato [...]

I've found that replacing files can be a bad idea because it can mess up extended scripts.

I was trying a few things with the file *character_selection.tscn*, and after replacing it I found that the extended scripts for *character_selection.gd* (which is referenced in *character_selection.tscn*) no longer worked. Replacing *.tres* files likely has the same negative effect, where any *.gd* scripts that are referenced in that *.tres* file will ignore their extensions.

This needs further testing though.
